hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -0,0 +1,302 @@
+from elftools.elf.sections import SymbolTableSection
+from tabulate import tabulate
+
+from .insn import JAL, JALR, LOOP, LOOPI
+from .isa import OTBNInsn
+from .state import OTBNState
+from .trace import Trace
+
+
+class ExecutionStats:
+    def __init__(self) -> None:
+        self.stall_count = 0
+        self.insn_count = 0
+        self.instruction_histo = Counter()
+        self.func_calls = [] # type: List[Dict[str, str]]
+        self.loops = [] # type: List[Dict[str, str]]
+
+    def record_stall(self) -> None:
+        self.stall_count += 1
+
+    def record_insn(self, pc: int, insn: OTBNInsn,
+                    state_before_commit: OTBNState) -> None:
+        '''Collect execution statistics'''
+
+        self.insn_count += 1
+
+        # Instruction histogram
+        self.instruction_histo[insn.insn.mnemonic] += 1
+
+        # Function calls
+        # - Direct function calls: jal x1, <offset>
+        # - Indirect function calls: jalr x1, <grs1>, 0
+        if isinstance(insn, JAL) and insn.grd == 1:
+            call_stack = state_before_commit.peek_call_stack()
+            if call_stack:
+                caller_func = call_stack[0]
+            else:
+                caller_func = state_before_commit.start_address
+
+            self.func_calls.append({
+                'call_site': pc,
+                'caller_func': caller_func,
+                'callee_func': state_before_commit.pc_next,
+            })
+
+        # Loops
+        if isinstance(insn, LOOPI):
+            self.loops.append({
+                'loop_addr': pc,
+                'loop_len': insn.bodysize,
+                'iterations': insn.iterations,
+            })
+
+        if isinstance(insn, LOOP):
+            iters = state_before_commit.gprs.get_reg(insn.grs).read_unsigned()
+            self.loops.append({
+                'loop_addr': pc,
+                'loop_len': insn.bodysize,
+                'iterations': iters,
+            })
+
+
+def _dwarf_decode_file_line(dwarf_info: DWARFInfo,
+                            address: int) -> Optional[Tuple[str, int]]:
+    # Go over all the line programs in the DWARF information, looking for
+    # one that describes the given address.
+    for CU in dwarf_info.iter_CUs():
+        # First, look at line programs to find the file/line for the address
+        lineprog = dwarf_info.line_program_for_CU(CU)
+        prevstate = None
+        for entry in lineprog.get_entries():
+            # We're interested in those entries where a new state is assigned
+            if entry.state is None:
+                continue
+            if entry.state.end_sequence:
+                # if the line number sequence ends, clear prevstate.
+                prevstate = None
+                continue
+            # Looking for a range of addresses in two consecutive states that
+            # contain the required address.
+            if prevstate and prevstate.address <= address < entry.state.address:
```

Comment:
Does this work properly if `address` equals `entry.state.address` for the last entry in `lineprog.get_entries()`?

Reply:
This and the other DWARF code is copied from a (public domain) example that is part of the Python libelf, I assume it works correctly.

Reply:
:-) I'm pretty certain that it wouldn't work if that happened. Maybe DWARF entries never look like that though.

Reply:
Yeah, there are too many oddities in DWARF, ELF, and friends for me to have any feeling of what happens -- certainly less than the author of this code :smile: I'll keep it in mind when I get back to DWARF parsing (I still want to figure out how to get good line numbers ...)

Reply:
Line numbers are HARD (there's a virtual machine involved). If we're lucky, the Python wrapper implements it already, but I've done it in C++ in the past and it took days to get right :-/

Reply:
We are able to get line numbers (you can compile an OTBN binary with -g and look at the stats produced in this PR) and they match what addr2line produces. So the Python code is probably good.

But the line numbers are not exactly what you'd expect them to be especially for functions which look like:

```
10:  last_insn_in_previous_label
11:  
12: /*
13:  * Comment about the next section
14:  */
15: subroutine_label:
16:   first_subroutine_insn
```

In my experiments, the line number for `first_subroutine_insn` (i.e. the address of `subroutine_label`) would be 11, where a developer would hope for 15 or 16.

I initially thought that we don't have the right `.line` statements in the transformed assembly, but they look correct when I did examine them more closely.

It's effectively up to the assembler to assign a semantically "helpful" line number to a statement, and I've played around with a couple of options in gas, it seems there are different numbers embedded (try different `-a` options). 

That's roughly where I stopped caring :-)

Reply:
I suspect that it's to do with otbn-as's pseudo-instruction expansion. Something like `li` may expand to 2 lines and I don't think I've got the right `.line` directives afterwards. If you have an easy example to debug, I'll happily take a look.

Reply:
I initially thought as well it's due to the 2-line expansion, but when looking further it didn't seem to be the problem ... I'll file a proper issue with example code after my vacation.